rfcs/0008-microtvm-project-api.md

```diff
@@ -0,0 +1,509 @@
+implementation of the **Project API** inside an executable program known as the
+**TVM Project API Server**.
+
+To begin working with a particular platform's Project API implementation, the user supplies TVM with
+the path to the top-level directory. TVM launches an instance the Project API Server (found at a
+standard location in that directory). TVM communicates with the Project API Server using JSON-RPC
+over standard OS pipes.
+
+TVM supplies generated code to the Project API Server using [Model Library Format](0001-model-library-format.md).
+
+Below is a survey of example workflows used with the Project API Server:
+
+## Generating a project
+
+1. The user imports a model into TVM and builds it using `tvm.relay.build`.
+2. The user supplies TVM with the path to the template project and a path to a non-existent
+   directory where the generated project should live.
+3. TVM launches a Project API server in the template project.
+4. TVM verifies that the template project is indeed a template by invoking the Project API server
+   `server_info_query` method.
+5. TVM invokes the Project API server `generate_project` method to generate the new project.
+
+## Building and Flashing
+
+1. The user follows the steps under [Generating a project](#generating-a-project).
+2. TVM expects the Project API server to copy itself to the generated project. It launches a
+   Project API server in the generated project directory.
+3. TVM verifies that the generated project is not a template by invoking the Project API server
+   `server_info_query` method. This method also returns `options` that can be used to customize
+   the build.
+4. TVM invokes the Project API server `build` method to build the project.
+5. TVM invokes the Project API server `flash` method to program the attached device. The
+   `options` can be used to specify a device serial number.
+
+## Host-driven model inference
+
+1. The user follows the steps under [Generating a project](#generating-a-project).
+2. TVM invokes the Project API server `connect_transport` method to connect to the remote on-device
+   microTVM RPC server.
+3. The microTVM RPC server is attached to a traditional TVM RPC session on the host device.
+4. TVM drives inference on-device using the traditional TVM RPC methods. The Project API server
+   methods `read_transport` and `write_transport` are used to receive and send data.
+5. When the inference session is over, TVM invokes the Project API server method `close_transport`
+   to release any underlying I/O resources, and terminates the Project API server.
+
+## AutoTVM
+
+1. The user supplies a kernel to the AutoTVM tuner for search-based optimization.
+2. AutoTVM generates a set of task configurations, instantiates each task, and then invokes
+   `tvm.build` to produce a `Module` for each instantiated task.
+3. AutoTVM produces a Model Library Format archive from the `Module` for each instantiated task.
```

Comment:
Just to clarify, this is one combined model library format with all `Module`s combined so it only flashes once for many task permutations?

Reply:
correct, though the number of tasks here is not specified. that still needs to be added to AutoTVM past the initial work done in https://github.com/apache/tvm/pull/7545